packer/plugin-getter/plugins.go

```diff
@@ -0,0 +1,421 @@
+			versionStr := strings.TrimPrefix(fname, filenamePrefix)
+			versionStr = strings.TrimSuffix(versionStr, filenameSuffix)
+			pv, err := version.NewVersion(versionStr)
+			if err != nil {
+				// could not be parsed, ignoring the file
+				log.Printf("found %q with an incorrect %q version, ignoring it. %v", path, versionStr, err)
+				continue
+			}
+
+			// no constraint means always pass, this will happen for implicit
+			// plugin requirements
+			if !pr.VersionConstraints.Check(pv) {
+				log.Printf("[TRACE] version %q of file %q does not match constraint %q", versionStr, path, pr.VersionConstraints.String())
+				continue
+			}
+
+			checksumOk := false
+			for _, checksummer := range opts.Checksummers {
+
+				cs, err := checksummer.GetChecksumOfFile(path)
+				if err != nil {
+					log.Printf("[TRACE] GetChecksumOfFile(%q) failed: %v", path, err)
+					continue
+				}
+
+				if err := checksummer.ChecksumFile(cs, path); err != nil {
+					log.Printf("[TRACE] ChecksumFile(%q) failed: %v", path, err)
+					continue
+				}
+				checksumOk = true
+				break
+			}
+			if !checksumOk {
+				log.Printf("[TRACE] No checksum found for %q ignoring possibly unsafe binary", path)
+				continue
+			}
+
+			res.InsertSortedUniq(&Installation{
+				BinaryPath: path,
+				Version:    versionStr,
+			})
+		}
+	}
+	return res, nil
+}
+
+// InstallList is a list of installs
+type InstallList []*Installation
+
+func (l InstallList) String() string {
+	v := &strings.Builder{}
+	v.Write([]byte("["))
+	for i, inst := range l {
+		if i > 0 {
+			v.Write([]byte(","))
+		}
+		fmt.Fprintf(v, "%v", *inst)
+	}
+	v.Write([]byte("]"))
+	return v.String()
+}
+
+// InsertSortedUniq inserts the installation in the right spot in the list by
+// comparing the version lexicographically.
+// A Duplicate version will replace any already present version.
+func (l *InstallList) InsertSortedUniq(install *Installation) {
+	pos := sort.Search(len(*l), func(i int) bool { return (*l)[i].Version >= install.Version })
+	if len(*l) > pos && (*l)[pos].Version == install.Version {
+		(*l)[pos] = install
+		return
+	}
+	(*l) = append((*l), nil)
+	copy((*l)[pos+1:], (*l)[pos:])
+	(*l)[pos] = install
+}
+
+// Installation describes a plugin installation
+type Installation struct {
+	// path to where binary is installed, if installed.
+	// Ex: /usr/azr/.packer.d/plugins/github.com/hashicorp/packer-plugin-amazon/packer-plugin-amazon_v1.2.3_darwin_amd64
+	BinaryPath string
+
+	// Version of this plugin, if installed and versionned. Ex:
+	//  * v1.2.3 for packer-plugin-amazon_v1.2.3_darwin_x5
+	//  * empty  for packer-plugin-amazon
+	Version string
+}
+
+// InstallOptions describes the possible options for installing the plugin that
+// fits the plugin Requirement.
+type InstallOptions struct {
+	// Different means to get releases, sha256 and binary files.
+	Getters []Getter
+
+	// Any downloaded binary and checksum file will be put in this folder.
+	InFolders []string
```

Comment:
if there are more than one folder defined, how do we choose what folder to put the downloaded binary into?